src/motor_hardware.cc

```diff
@@ -863,53 +864,52 @@ void MotorDiagnostics::firmware_options_status(DiagnosticStatusWrapper &stat) {
 // The I2C address is the 8-bit address which is the 7-bit addr shifted left in some code
 // If chipRegAddr is greater than 1 we write this out for the internal chip address for the following read(s)
 //
-// Returns number of bytes read where 0 or less implies some form of failure
-//
 // NOTE: The i2c8bitAddr will be shifted right one bit to use as 7-bit I2C addr
 //
 static int i2c_BufferRead(const char *i2cDevFile, uint8_t i2c8bitAddr, 
-                          uint8_t *pBuffer, int16_t chipRegAddr, uint16_t NumByteToRead)
+                          uint8_t *pBuffer, int16_t chipRegAddr)
 {
-   int bytesRead = 0;
-   int retCode   = 0;
-
-    int fd;                                         // File descrition
-    int  address   = i2c8bitAddr >> 1;              // Address of the I2C device
-    uint8_t buf[8];                                 // Buffer for data being written to the i2c device
+    int fd;                                         // File descriptor
+    int retCode = 0;
+    int slaveAddress = i2c8bitAddr >> 1;            // Address of the I2C device
 
     if ((fd = open(i2cDevFile, O_RDWR)) < 0) {      // Open port for reading and writing
-      retCode = -2;
       ROS_ERROR("Cannot open I2C def of %s with error %s", i2cDevFile, strerror(errno));
+      retCode = -1;
       goto exitWithNoClose;
     }
 
-    // The ioctl here will address the I2C slave device making it ready for 1 or more other bytes
-    if (ioctl(fd, I2C_SLAVE, address) != 0) {        // Set the port options and addr of the dev
-      retCode = -3;
-      ROS_ERROR("Failed to get bus access to I2C device %s!  ERROR: %s", i2cDevFile, strerror(errno));
-      goto exitWithFileClose;
-    }
+    uint8_t buf[1];                                 // Buffer to be written to i2c device 
+    uint8_t outbuf[1];                              // Buffer to be read into from i2c device
+    struct i2c_msg msgs[2];
+    struct i2c_rdwr_ioctl_data msgset[1];
+    msgs[0].addr = slaveAddress;                    // Slave adress
+    msgs[0].flags = 0;                              // Write bit
+    msgs[0].len = 1;                                // Number of data bytes written to I2C slave address 
+    msgs[0].buf = buf;                              // Writting to register
```

Comment:
Does this ioctl  and  msgs[0].len = 1 write to the device ?
If so that is a problem.  Writing to this device configures pins to be inputs or outputs.
We should not write to this  device over I2C.

Reply:
That's how the I2C protocol is defined. In order to read from the i2c device register, you must first tell it from which of its internal addresses you want to read, which is done by writing to it. I think there is no way to actually immediatelly read from the device. The only problem I see it might occur is clock streching(which might be automatic on our i2c devices and the problem is gone in that case).

Reply:
The reason I ask is that the PCF8574 is a very simple device and does not have internal chip addressing for registers.  There is only one register to write and that confures the pins as inputs or outputs.  We do not want to write to this chip except to configure it for all inputs in this case.  Find TI datasheet for PCF8574 and you will find this usage info:

This device does not have internal configuration or
status registers. Instead, read or write to the device I/Os directly after sending the device address

That is why the interface to this read routine had a special trap for chipRegAddr so it could suppress the writing of the 'register address' to the chip which in this case would re-configure the chip to all outputs if we wrote out a 0x00.  If we write to this chip at all it should be 0xff for all inputs BUT we do not want to write to it at all or general usage of the routine could not deal with a PCF8574 we use in a different way.   I have 3 pins on this chip that could be outputs because they go to a jack so the routine to access the part must not make assumptions on chip configuration.